File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum $ filter isMultiple [1..(limit-1)]
    where
    isMultiple n = any (factorOf n) factors
    factorOf n = (0 ==) . mod n

File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sum(factors: Set[Int], limit: Int): Int = {
    var total = 0
    for (i <- 1 to limit - 1) {
      if (isMultiple(factors, i)) {
        total += i
      }
    }
    total
  }

  def isMultiple(factors: Set[Int], candidate: Int) = {
    var result = false
    for (fact <- factors) {
      if (candidate % fact == 0) {
        result = true
      }
    }
    result
  }
}

